built-in fonts and custom fonts examples

alexvoss opened the pull request:
Pull request for fonts examples
https://github.com/mkdocs-material/internal/issues/29

alexvoss committed:
added example for using only built-in fonts

squidfunk commented on examples/fonts-builtin/docs/README.md:
```diff
@@ -0,0 +1,45 @@
+Material for MkDocs provides sensible defaults that should work in all
+common browsers. 
+
+## (Dis-)Advantages
```
I'd say we should mention the privacy plugin to have a "best of both worlds" solution ☺️ Not sure where, but a small admonition that explains the plugin (just take one from our docs) might be good

alexvoss replied:
Add the tip - doing the same in the custom fonts example I just pushed a draft of.

alexvoss committed:
added tip pointing to privacy plugin

alexvoss committed:
initial draft of example using custom fonts

alexvoss commented:
I am trying to find a CDN for Noto (or another font really) that is not Google and that is not like Google. I have put in cdnfonts.com for now but I can't seem to find much info on who they are, how they make their money, etc. Do you know anything about jsDeliver? That seems to be the default one for Noto.

alexvoss committed:
+configuring fonts with your own CSS

alexvoss commented:
@squidfunk I will need to go for a walk (rare sunshine, so will head to the beach) and do a little shopping. Any feedback on this very welcome. I hope I can finish off any remaining work in the afternoon.

squidfunk commented on examples/fonts-builtin/docs/README.md:
```diff
@@ -0,0 +1,55 @@
+# Using built-in browser fonts
+
+This example shows what happens when auto-loading of fonts from Google
+is turned off and not other fonts are configured:
```
Typo "not" -> "no"

squidfunk commented on examples/fonts-builtin/docs/README.md:
```diff
@@ -0,0 +1,55 @@
+!!! tip "Privacy Plugin"
+
+    The [privacy plugin] provided by Material for MkDocs provides a
+    best-of-both-worlds solution in that it allows you to specifc 
```
Typo "specifc" -> "specify"

squidfunk commented on examples/fonts-custom/docs/README.md:
````diff
@@ -0,0 +1,145 @@
+    }
+    ```
+
+### Hosting on your own server
````
I'd just call this "alongside your documentatin" or something like it. "Server" might be off-putting, because it sounds like you need to set up hosting yourself. Just put them in docs, deploy to GitHub pages, done.

squidfunk commented on examples/fonts-custom/docs/README.md:
```diff
@@ -0,0 +1,145 @@
+If you are loading the fonts from a CDN as in this example, you are
+still not limiting traffic to only your website. 
+
+TODO: what about fonts hosted on GitHub Pages? A lot of documentation lies on GHP anyway, so why not load the fonts from there? Do they offer this?
```
See above

squidfunk commented on examples/fonts-custom/docs/README.md:
````diff
@@ -0,0 +1,145 @@
+
+=== "`extra.css`"
+
+    ```css
````
Why not use Snippets here? Less duplication and the code example will always be up to date ☺️

alexvoss replied:
Snippets need to be added to `mkdocs.yml` and so that is no longer minimal. Already had that with admonitions.

squidfunk reviewed (changes requested):
I'm currently busy working on the search, so I didn't run the PR, only minor changes. If it works and you feel comfortable deploying it, I'm fine with merging it and linking it in the docs.

alexvoss committed:
put font spec into the line: family, weight, style

alexvoss committed:
fixed typos

alexvoss committed:
added locally hosted font stylesheet

alexvoss committed:
fixed typo; added to (dis-)advantages section

alexvoss commented:
Right, should be fine now. Fingers crossed. I did test that both examples work with both insider and public version.

alexvoss committed:
activated the CDN version again so we do not need to host fonts

squidfunk commented on examples/fonts-custom/docs/README.md:
```diff
@@ -0,0 +1,196 @@
+
+## (Dis-)Advantages
+
+If you are loading the fonts from a CDN as in this example, you are
```
One final question: what line gutters are you using? I'm trying to limit everything to 80 chars, as it allows to render code blocks in PDFs nicely, but it looks like you're using 70 or something in that area. Could we agree on 80? Sorry to be so pendantic 😅

alexvoss replied:
Indeed, it seems that for some reason it is 70 on `.md` files. This is `vim`'s doing. I have changed it to 80. Shame this does not seem to be something that a `.editorconfig` could take care of?! Do you want me to change these or just use 80 from here onwards?

alexvoss replied:
I am also just realizing that I do not understand the comment about code blocks. Those sometimes get longer, especially when URLs are involved. I am not sure what to do about that.

You can stop apologizing for being "pedantic" - that is quite ok and we know each other well enough by now!

squidfunk replied:
> Indeed, it seems that for some reason it is 70 on .md files. This is vim's doing. I have changed it to 80. Shame this does not seem to be something that a .editorconfig could take care of?! Do you want me to change these or just use 80 from here onwards?

Please just use 80 onwards. When we revisit those docs at some point, we can correct them the next time.

> I am also just realizing that I do not understand the comment about code blocks. Those sometimes get longer, especially when URLs are involved. I am not sure what to do about that.

When you print a site made with Material for MkDocs from the browser, limiting code blocks to 80 characters will always nicely fit the page. That's one of the reasons I always stick to the 80 characters rule. Additionally, it allows to open multiple editor tabs next to each other without having to scroll horizontally. I'm just not that big a fan of horizontal scrolling on text ☺️ So to sum up: 80 is just the best practice I've been following for many years.

> You can stop apologizing for being "pedantic" - that is quite ok and we know each other well enough by now!

Ack!

alexvoss replied:
I don't question the 80 chars rule (I had 70 configured, probably on the basis of recommended line length for reading), just wondering what to do with things like this in the CSS file:

```
https://cdn.jsdelivr.net/gh/notofonts/notofonts.github.io/fonts/NotoSerif/unhinted/otf/NotoSerif-Bold.otf
```

alexvoss commented:
Not sure where we are with this - I thought this had runs its course but GitHub says there are change request to be addressed? I will ask you to review again, mainly because I don't know what else to do.

alexvoss commented:
@squidfunk one example that is missing is that of auto-loading another font from Google Fonts. That does raise the question of data protection and whether we should not use the privacy plugin here. I do note that Google Analytics is turned on, so would we not also need to get peoples' consent?

squidfunk commented:
Hey, sorry that I'm sometimes unresponsive right now, I'm focusing on the search, so I try to minimize distractions, which is why some things slip.

> That does raise the question of data protection and whether we should not use the privacy plugin here.

We can use the plugin, but that would make the example impossible to run without Insiders. Note that Google Analytics is configured in `mkdocs.yml`, but no `GOOGLE_ANALYTICS_KEY` variable is set during the build, so the integration is not active. It's only Google Fonts that is loaded.

I'd say we keep it simple.

__Edit__: you're also [using CDNs for the non-local example](https://github.com/mkdocs-material/examples/pull/11/files#diff-d666f757ba97a9d4483a63a17e77fa725eaffe23984826f4a8f012dd3b5283fb), which would mandate for the same logic. I take the risk and get sued over using a CDN before getting consent for every third-party request in any example. Adding the cookie consent to each of them defies the meaning of examples.

squidfunk commented on examples/fonts-builtin/mkdocs.yml:
```diff
@@ -0,0 +1,44 @@
+# -----------------------------------------------------------------------------
+
+# Project information
+site_name: Using built-in broswer fonts
```
Typo `broswer` 😉 Did you check for typos?

alexvoss replied:
another config change for my `.vimrc`, so should be better in the future.

squidfunk commented on examples/fonts-custom/docs/assets/stylesheets/extra.css:
```diff
@@ -0,0 +1,33 @@
+@font-face {
+  font-family: "Noto Serif"; font-weight: normal; font-style: normal;
```
Please break those onto separate lines (other CSS files as well)

squidfunk commented on examples/fonts-custom/docs/README.md:
````diff
@@ -0,0 +1,196 @@
+    }
+    ```
+
+### Using a CDN with your own CSS
````
I'm not sure we need this example – it's actually only the first step for self hosting, as the font files are still hosted on the CDN. I think I would drop this one and recommend the privacy plugin which takes care of all of this. With the privacy plugin, you get self hosting with the first example and a single line of config. It will be released when we reach the next funding goal, which is just around the corner.

alexvoss committed:
fixed typo

alexvoss commented:
> We can use the plugin, but that would make the example impossible to run without Insiders.

We could use the privacy plugin for building the whole set of examples and apply this to the individual examples through the transformation script. That would mean the deployed version would be GDPR compliant and the downloadable examples would not assume the use of Insiders.

alexvoss committed:
removed CDN with own CSS example; fixed local host example

and also fixed couple of typos/grammar errors

alexvoss committed:
put each CSS attribute into its own line

alexvoss commented:
Ok, I think I have addressed the requested changes. Also moved the mention of the privacy plugin closer to the top of the page on the custom fonts example.

squidfunk commented:
LGTM! By the way, once a PR is approved on this repository, feel free to merge it whenever you feel it's ready. On the actual Material for MkDocs repository, we need to carefully weigh when PRs can be merged, because it impacts a lot of users, but this repository is for examples/demonstration only, so we can relax it a bit.
